torch/csrc/autograd/engine.cpp

```diff
@@ -760,47 +786,121 @@ auto Engine::execute(const edge_list& roots,
     graph_task->init_to_execute(*graph_root, outputs);
   }
 
-  return execute_with_graph_task(graph_task, graph_root)->wait();
+  variable_list results = execute_with_graph_task(graph_task, graph_root)->wait();
+  // If it's owning thread backward call (not reentrant), local_ready_queue must
+  // be empty when we finish Engine::execute
+  if (not_reentrant_backward_call) {
+    TORCH_INTERNAL_ASSERT(local_ready_queue->empty());
+  }
+  return results;
 }
 
-void Engine::initialize_threads_pool() {
+void Engine::initialize_device_threads_pool() {
   track_bad_autograd_forks();
   TORCH_CHECK(!in_bad_autograd_fork,
               "Unable to handle autograd's threading in combination with fork-based multiprocessing. "
               "See https://github.com/pytorch/pytorch/wiki/Autograd-and-Fork");
-  std::call_once(start_threads_flag_, &Engine::start_threads, this);
+  std::call_once(start_device_threads_flag_, &Engine::start_device_threads, this);
 }
 
 void Engine::enqueue_blocked_task_on_cpu(NodeTask task) {
-  initialize_threads_pool();
-  ready_queue(at::kCPU).push(
+  initialize_device_threads_pool();
+  std::shared_ptr<GraphTask> graph_task = task.base_.lock();
+  // The graph_task must be alive at this point, because internal autograd machinary
+  // who calls this API (Distributed Autograd Engine) increases outstanding_tasks_
+  // outside this API to keep the GraphTask alive
+  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!");
+  ready_queue(graph_task, at::kCPU)->push(
       std::move(task), /* incrementOutstandingTasks */ false);
 }
 
+void Engine::execute_graph_task_with_continuation(const std::shared_ptr<GraphTask>& graph_task) {
+  std::shared_ptr<ReadyQueue> graph_task_rq = graph_task->cpu_ready_queue_;
+  while(!graph_task_rq->empty()) {
+    NodeTask task = graph_task_rq->pop();
+    std::shared_ptr<GraphTask> local_graph_task;
+    if (!(local_graph_task = task.base_.lock())) {
+      continue;
+    }
+    {
+      if (task.fn_ && !local_graph_task->has_error_.load()) {
+        AutoGradMode grad_mode(local_graph_task->grad_mode_);
+        try {
+          evaluate_function(local_graph_task, task.fn_.get(), task.inputs_);
+        } catch (std::exception& e) {
+          thread_on_exception(local_graph_task, task.fn_, e);
+          // early return in error so that we immediately stop the execution
+          // of this GraphTask and return the future with proper ErrorMessage
+          return;
+        }
+      }
+    }
+    // Decrement the outstanding task.
+    --local_graph_task->outstanding_tasks_;
+  }
+  // Check if we've completed execution.
+  if (graph_task_completed(graph_task)) {
+    // We don't need to explicitly notify the owner thread, since
+    // 'mark_graph_task_completed' would mark the Future as completed and this
+    // would notify the owner thread that the task has been completed.
+    mark_graph_task_completed(graph_task);
+  } else {
+    // schedule a continuation
```

Comment:
It seems we are scheduling continuation after evaluating every node. Can we do continuation when ready queue is empty? Since current PR perf is reasonable, we can do that as a follow up experiment.

Reply:
Never mind. I read the code wrong. We are doing continuation when queue is empty.